explorer/db/hooks_test.go

```diff
@@ -0,0 +1,153 @@
+package db_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/omni-network/omni/explorer/db"
+	"github.com/omni-network/omni/explorer/db/ent"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type results struct {
+	blocks   []ent.Block
+	messages []*ent.Msg
+	receipts []*ent.Receipt
+}
+
+type prerequisite func(t *testing.T, ctx context.Context) results
+
+func TestMsgHooks(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	type want struct {
+		BlockCount        int
+		MsgCount          int
+		ReceiptCount      int
+		StrayReceiptCount int
+		StrayMessageCount int
+	}
+
+	tests := []struct {
+		name          string
+		prerequisites prerequisite // These functions create entries on our db before the evaluation
+		wantErr       bool
+		want          want
+	}{
+		{
+			name: "create_initial_block_with_msg_then_following_contain_receipts",
+			prerequisites: func(t *testing.T, ctx context.Context) results {
+				t.Helper()
+				client := db.CreateTestEntClient(t)
+				t.Cleanup(func() {
+					if err := client.Close(); err != nil {
+						t.Error(err)
+					}
+				})
+				blocks := db.CreateTestBlocks(t, ctx, client, 2)
+
+				var messages []*ent.Msg
+				var receipts []*ent.Receipt
+				for _, b := range blocks {
+					messages = append(messages, b.QueryMsgs().AllX(ctx)...)
+					receipts = append(receipts, b.QueryReceipts().AllX(ctx)...)
+				}
+
+				return results{
+					blocks:   blocks,
+					messages: messages,
+					receipts: receipts,
+				}
+			},
+			want: want{
+				BlockCount:        2,
+				MsgCount:          2,
+				ReceiptCount:      1,
+				StrayReceiptCount: 0,
+				StrayMessageCount: 1,
+			},
+			wantErr: false,
+		},
+		{
+			name: "create_block_then_receipt_then_msg",
+			prerequisites: func(t *testing.T, ctx context.Context) results {
+				t.Helper()
+				client := db.CreateTestEntClient(t)
+				t.Cleanup(func() {
+					if err := client.Close(); err != nil {
+						t.Error(err)
+					}
+				})
```

Comment:
I believe that this entire block:
```go
client := db.CreateTestEntClient(t)
t.Cleanup(func() {
	if err := client.Close(); err != nil {
		t.Error(err)
	}
})
```
 has be moved inside a helper function `client := setupDB(t)` because you repeat that in every test. Then the test prerequisite would just receive a `client` argument.

Reply:
I ran into table doesn't exist errors when running the tests all at once in parellel so I moved it into the functions, however I just tested with calling a `setupDB(t)` method in each of the prerequisites, and that did work

Reply:
good catch!